### ISSUE TITLE
Allow HTTP 200 response for HTTP file upload PUT request

### ISSUE THREAD

**Issue by lorenz** [tigase/siskin-im#231]
**Is your feature request related to a problem? Please describe.**
Allow Siskin to use HTTP file upload with backend services returning HTTP 200 for new files. For example all object stores based on the widely-used Amazon S3 API do this.

**Describe the solution you'd like**
Accept HTTP 200 as a valid response for the HTTP PUT request.

Something like this should take care of it as all non 200/201 status codes are already filtered out.
```diff
- if code == 200 {
-   completionHandler(.failure(.invalidResponseCode(url: slot.getUri)));
- } else {
  completionHandler(.success(slot.getUri));
- }
```

**Describe alternatives you've considered**
Otherwise it's not possible to directly use object stores and all PUTs need to be forcibly redirected through a proxy which is annoying.

**Additional context**
Most other clients (Conversations, Dino, ...) are fine with this, just Siskin doesn't like it.

**Comment by licaon-kter** [comment]
Isn't thisa server feature? Like ejabberd has a s3 contrib module, maybe that module says 201 instead? 

How are you storing the files? Which xmpp server software?

**Comment by lorenz** [comment]
Yes, this is a server feature. But the thing is that the thing that's serving the HTTP upload requests is not the XMPP server, but instead an object storage directly. The "files" are objects in a Minio S3 cluster and the upload goes straight to that (the XMPP server only provides the presigned URL serving as an authorization). This is with Prosody and mod_http_upload_s3. ejabberd with S3 should IMO have the same problem, do we know that that works?

**Comment by hantu85** [comment]
Siskin is closely following specification of XEP-0363: HTTP File Upload, see https://github.com/tigase/siskin-im/issues/37#issuecomment-624294330 for the details as to why Siskin shows warning dialog.

**Comment by arturhefczyc** [comment]
@lorenz the best and the correct way to ensure all software works together and is compatible, is when everybody follows the spec. Suggesting to deviate from the spec to be compatible with others who deviated from the spec is counterproductive. If we follow this path, we end up with chaos.

If the spec is wrong, change the spec and let everybody adjust software to adhere to the spec.

**Comment by raucao** [comment]
+1 for supporting 200. The spec shouldn't even require 201 specifically, but leave that to the HTTP spec IMO. 200 is just as good as 201 across almost all of the Web. It's an easy fix here, and we have the same issue with the ejabberd module indeed. Just switched to it a couple of days ago, and all clients I tested work fine, except for only the Apple ones (Beagle, Siskin, Snikket, Monal).

**Comment by lorenz** [comment]
I would indeed prefer if S3 returned 201 (for POST this is even configurable with `success_action_status`) as that would be the more correct response. But that ship has sailed more than 10 years ago, S3 is THE object storage API (pretty much any object storage system implements it). It is basically impossible to get AWS to change their implementation (as it's a breaking change) and thus all third-party implementations won't either.

I contacted the author of the XEP to maybe update it or issue an errata, but haven't heard back.

**Comment by raucao** [comment]
@hantu85 Yeah, I just meant the spec doesn't have to include that at all, since it's already in the HTTP spec. But even then, half the Web is violating that spec for PUTs anyway, and as @lorenz said, we won't get the S3 API to change either.

Btw, here's how I fixed it for our API, using the [Nginx Lua Module](https://www.nginx.com/resources/wiki/modules/lua/): https://gist.github.com/raucao/0aa1c59a1b3902b2a7a7ff925f0c32d4#file-nginx-s3-vhost-conf-L14-L18
